### PR TITLE
fix(cli/installer): Strip "@..." suffixes from inferred names

### LINKED ISSUE #7221: Invalid Executable Name
See [here](https://github.com/drashland/dmm/issues/61) for reference, where this problem originated from.

I can also re-produce the error.

**What is the Problem**
Getting a `invalid executable name` when installing a module

**What is Expected**
Can install the module

**How to Reproduce**
1. run: `deno install --allow-net --allow-read --allow-write https://deno.land/x/dmm@v1.1.5/mod.ts`
2. see error: `invalid executable name: dmm@v1.1.5`
3. pass in the `--name` flag, and it works. Although in the above link, someone is unable to install even when using the `--name`

### PR DESCRIPTION
**Pull request by nayeemrmn**
Fixes #7221.

### CODE REVIEW

**Comment by bartlomieju** [comment]
@nayeemrmn there are more characters that are not permitted for filenames; eg. `:` on Windows - this boils down to similar problem as in `$DENO_DIR` cache keys.

**Comment by nayeemrmn** [comment]
> there are more characters that are not permitted for filenames; eg. `:` on Windows - this boils down to similar problem as in `$DENO_DIR` cache keys.

We can't always infer a valid name. If someone wants to put `:` in their executable module name, it's fine if it throws `Invalid executable name` when installed with an implicit name so they're encouraged to name things differently. The purpose of this PR is to accommodate the `@<tag>` convention since we know how to unpack it.

File: cli/installer.rs [hunk 1]
```diff
@@ -349,6 +348,24 @@ mod tests {
...
+      infer_name_from_url(
+        &Url::parse("https://example.com/abc@0.1.0").unwrap()
+      ),
+      Some("abc".to_string())
+    );
+    assert_eq!(
+      infer_name_from_url(
+        &Url::parse("https://example.com/abc@0.1.0/main.ts").unwrap()
+      ),
+      Some("abc".to_string())
```
Comment (ry) [inline cli/installer.rs:361]: I didn't even realize we already supported this fancy inference:
```
    assert_eq!(
      infer_name_from_url(&Url::parse("file:///abc/main.ts").unwrap()),
      Some("abc".to_string())
    );
```
I suppose given that this patch is useful.

Comment (ry) [reply cli/installer.rs:361]: I feel a little queazy about the fancy inference tho. It's one of these magical algorithms that is just another thing to remember.

**Pull request merged**